qiskit/dagcircuit/dagcircuit.py

```diff
@@ -1371,6 +1371,36 @@ def substitute_node(self, node, op, inplace=False):
             self._decrement_op(node.op)
         return new_node
 
+    def swap_nodes(self, node1, node2):
+        """Swap connected nodes e.g. due to commutation.
+
+        Args:
+            node1 (OpNode): predecessor node
+            node2 (OpNode): successor node
+
+        Raises:
+            DAGCircuitError: if either node is not an OpNode
+        """
+        if not (isinstance(node1, DAGOpNode) and isinstance(node2, DAGOpNode)):
+            raise DAGCircuitError("nodes to swap are not both OpNodes")
+        try:
+            connected_edges = self._multi_graph.get_all_edge_data(node1._node_id, node2._node_id)
+        except rx.NoEdgeBetweenNodes:
+            # since nodes aren't connected there's nothing to do
+            return
+        node1_id = node1._node_id
+        node2_id = node2._node_id
+        edge_find = lambda x: x == edge
+        for edge in connected_edges[::-1]:
+            edge_parent = self._multi_graph.find_predecessors_by_edge(node1_id, edge_find)[0]
```

Comment:
The delayed closure of this `lambda` is _very_ confusing.  At the time the `lambda` is written, the variable `edge` would return `UnboundLocalException` on access.  It will only succeed later, within the loop, once `edge` has been assigned to for the first time.

I'm surprised pylint isn't complaining quite a lot about this for that reason, and because it usually doesn't like direct assignment of lambdas.

Reply:
I added `edge` as a default argument to the lambda which I understand avoids the delayed closure.